### PR TITLE
feat(builder, server): hide secrets from the UI and mark them as secrets in the api

### PR DESCRIPTION
**Pull request by ntindle**
### Background

If you screenshot or record the builder, your API keys will be leaked. This change is designed to ensure that does not happen.

### Changes 🏗️

- Adds a launch.json for debugging the frontend
- Adds a parameter to the API to mark something secret
- Adds a toggle to the frontend to show something is secret

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [ ] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [ ] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [ ] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [ ] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added `secret` parameter to `SchemaField` and `SecretField` functions in the backend to mark fields as secret.
- Modified the `CustomNode` component in the frontend to handle and display secret fields as masked.
- Added a `launch.json` configuration file for debugging Next.js applications, including server-side, client-side, and full-stack debugging options.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>model.py</strong><dd><code>Add secret parameter to SchemaField and SecretField functions</code></dd></summary>
<hr>

rnd/autogpt_server/autogpt_server/data/model.py

<li>Added <code>secret</code> parameter to <code>SchemaField</code> and <code>SecretField</code> functions.<br> <li> Included logic to handle <code>secret</code> parameter in <code>json_extra</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7490/files#diff-2bb846d80ae85f951ecd4beb9837cc39b112b6a1970e2f90cf95b29f987925af">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CustomNode.tsx</strong><dd><code>Implement secret field handling in CustomNode component</code>&nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_builder/src/components/CustomNode.tsx

<li>Modified <code>renderClickableInput</code> to handle secret fields.<br> <li> Updated input rendering logic to display secret fields as masked.<br> <li> Added conditional checks for <code>schema.secret</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7490/files#diff-ce37ba33e874cc42255fda0cd2a16758a94a6e68606d513b7018c1cc68193146">+48/-26</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>launch.json</strong><dd><code>Add launch configurations for Next.js debugging</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_builder/.vscode/launch.json

<li>Added launch configurations for debugging Next.js applications.<br> <li> Included configurations for server-side, client-side, and full-stack <br>debugging.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7490/files#diff-528e2e75abb5cf4c430a2721063ab9e57d0b405efd0bcfa59c2a81b20cf1850c">+30/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by Torantulino** [comment]
Tested and working:
image

File: rnd/autogpt_builder/.vscode/launch.json
Comment (Torantulino) [inline rnd/autogpt_builder/.vscode/launch.json:1]: What is this?

**Review by Torantulino — APPROVED** [review]

File: rnd/autogpt_builder/.vscode/launch.json
Comment (ntindle) [reply rnd/autogpt_builder/.vscode/launch.json:1]: Launch debugger for the next JS stuff

Comment (Pwuts) [reply rnd/autogpt_builder/.vscode/launch.json:1]: out of scope! grr

Comment (ntindle) [reply rnd/autogpt_builder/.vscode/launch.json:1]: we all have our dirty habits ✨

**Pull request merged**